src/cryptonote_core/service_node_swarm.cpp

```diff
@@ -0,0 +1,272 @@
+  }
+
+  prod_static void get_excess_pool(size_t threshold, swarm_snode_map_t& swarm_to_snodes, std::vector<excess_pool_snode>& pool_snodes, size_t& excess)
+  {
+    /// Create a pool of all the service nodes belonging
+    /// to the swarms that have excess. That way we naturally
+    /// make the chances of picking a swarm proportionate to the
+    /// swarm size.
+    pool_snodes.clear();
+
+    if (threshold < MIN_SWARM_SIZE)
+      return;
+
+    excess = 0;
+    for (const auto &entry : swarm_to_snodes)
+    {
+      if (entry.second.size() > threshold)
+      {
+        excess += entry.second.size() - MIN_SWARM_SIZE;
+        for (const auto& sn_pk : entry.second)
+        {
+          pool_snodes.push_back({sn_pk, entry.first});
+        }
+      }
+    }
+  }
+
+  prod_static void create_new_swarm_from_excess(swarm_snode_map_t &swarm_to_snodes, std::mt19937_64 &mt)
+  {
+    while (calc_excess(swarm_to_snodes) >= calc_threshold(swarm_to_snodes))
+    {
+      LOG_PRINT_L2("New swarm creation");
+      std::vector<crypto::public_key> new_swarm_snodes;
+
+      while (new_swarm_snodes.size() < NEW_SWARM_SIZE)
+      {
+        std::vector<excess_pool_snode> pool_snodes;
+        size_t excess;
+        get_excess_pool(EXCESS_BASE, swarm_to_snodes, pool_snodes, excess);
+        if (pool_snodes.size() == 0)
+        {
+          MERROR("Error while getting excess pool for new swarm creation");
+          return;
+        }
+        const auto& random_excess_snode = pick_from_excess_pool(pool_snodes, mt);
+        new_swarm_snodes.push_back(random_excess_snode.public_key);
+        remove_excess_snode_from_swarm(random_excess_snode, swarm_to_snodes);
+      }
+      const auto new_swarm_id = get_new_swarm_id(mt, swarm_to_snodes);
+      swarm_to_snodes.insert({new_swarm_id, std::move(new_swarm_snodes)});
+      LOG_PRINT_L2("Created new swarm from excess: " << new_swarm_id);
+    }
+  }
+
+  prod_static void calc_swarm_sizes(const swarm_snode_map_t &swarm_to_snodes, std::vector<swarm_size> &sorted_swarm_sizes)
+  {
+    sorted_swarm_sizes.clear();
+    for (const auto &entry : swarm_to_snodes)
+    {
+      sorted_swarm_sizes.push_back({entry.first, entry.second.size()});
+    }
+    std::sort(sorted_swarm_sizes.begin(),
+              sorted_swarm_sizes.end(),
+              [](const swarm_size &a, const swarm_size &b) {
+                return a.size < b.size;
+              });
+  }
+
+  /// Assign each snode from snode_pubkeys into the FILL_SWARM_LOWER_PERCENTILE percentile of swarms
+  /// and run the excess/threshold logic after each assignment to ensure new swarms are generated when required.
+  prod_static void assign_snodes(const std::vector<crypto::public_key> &snode_pubkeys, swarm_snode_map_t &swarm_to_snodes, std::mt19937_64 &mt)
+  {
+    std::vector<swarm_size> sorted_swarm_sizes;
+    for (const auto &sn_pk : snode_pubkeys)
+    {
+      calc_swarm_sizes(swarm_to_snodes, sorted_swarm_sizes);
```

Comment:
Seems wasteful to have to call this in every iteration of the loop when we only change a single value.

Reply:
Resolving this as discussed in the office.